### PR TITLE
Updates LR schedule HOWTO for optax

### PR DESCRIPTION
**Pull request by matthias-wright**
This PR addresses #1577.

Note that I replaced the triangular scheduler with the cosine scheduler from the [imagenet example](https://github.com/google/flax/blob/cb5ebcd3b9ab8db10c79e88fcdb2dc67ce2668e2/examples/imagenet/train.py#L88) because the triangular scheduler did not play well with ``jit``. I think ``jit`` cannot trace the ``create_triangular_schedule`` function because of the if else that depends on the input.

### CODE REVIEW

File: docs/howtos/lr_schedule.rst [hunk 1]
```diff
@@ -1,138 +1,158 @@
...
-        lr = lr_max - ((cycle_step - top)/top) * (lr_max - lr_min)
-      return lr
-    return learning_rate_fn
-
-
-To use the schedule, one must create a learning rate function by passing the hyperparameters to the
-create_triangular_schedule function and then use that function to compute the learning rate for your updates.
-For example using this schedule on MNIST would require changing the train_step function
+  def create_learning_rate_fn(config, base_learning_rate, steps_per_epoch):
+    """Create learning rate schedule."""
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:30]: nit: use descriptive-style in function pydoc:
https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: docs/howtos/lr_schedule.rst [hunk 2]
```diff
@@ -1,138 +1,158 @@
...
+        init_value=base_learning_rate,
+        decay_steps=cosine_epochs * steps_per_epoch)
+    schedule_fn = optax.join_schedules(
+        schedules=[warmup_fn, cosine_fn],
+        boundaries=[config.warmup_epochs * steps_per_epoch])
+    return schedule_fn
+
+To use the schedule, we must create a learning rate function by passing the hyperparameters to the
+create_learning_rate_fn function and then pass the function to your ``optax`` optimizer.
+For example using this schedule on MNIST would require changing the train_step function:
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:45]: nit: ``train_step``

File: docs/howtos/lr_schedule.rst [hunk 3]
```diff
@@ -1,138 +1,158 @@
...
+    cosine_fn = optax.cosine_decay_schedule(
+        init_value=base_learning_rate,
+        decay_steps=cosine_epochs * steps_per_epoch)
+    schedule_fn = optax.join_schedules(
+        schedules=[warmup_fn, cosine_fn],
+        boundaries=[config.warmup_epochs * steps_per_epoch])
+    return schedule_fn
+
+To use the schedule, we must create a learning rate function by passing the hyperparameters to the
+create_learning_rate_fn function and then pass the function to your ``optax`` optimizer.
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:44]: nit: ``create_learning_rate_fn``

File: docs/howtos/lr_schedule.rst [hunk 4]
```diff
@@ -1,138 +1,158 @@
...
-    optimizer = optimizer.apply_gradient(grad, {"learning_rate": lr}) #!
+    (_, logits), grads = grad_fn(state.params)
+    new_state = state.apply_gradients(grads=grads)
     metrics = compute_metrics(logits, batch['label'])
-    return optimizer, metrics
+    lr = learning_rate_fn(state.step) #!
+    metrics['learning_rate'] = lr #!
+    return new_state, metrics
 
 And the train_epoch function:
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:81]: nit: ``train_epoch``

File: docs/howtos/lr_schedule.rst [hunk 5]
```diff
@@ -1,138 +1,158 @@
...
-
+  
     logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
-                epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)
+                 epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)
+  
+    return state, epoch_metrics
+
+
+And the create_train_state function:
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:136]: nit: ``create_train_state``

File: docs/howtos/lr_schedule.rst (hunk 3 above)
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:44]: consider adding a link to the Optax docs somewhere:
https://optax.readthedocs.io/en/latest/api.html#optimizer-schedules

**Review by andsteing — CHANGES_REQUESTED** [review]
So it seems the former HOWTO didn't actually work?
(Since the if..else problem is not something that has to do with Optax but rather a limitation of JIT)

This kind of proves the point that we should test these functions via doctest. Could you add some test code that covers the functions?

(sometimes we also link Colabs as starting points like in the `howtos/optax_update_guide.rst` but that's probably not needed here since this is rather short)

**Comment by matthias-wright** [comment]
Thanks for the feedback! No, it didn't work, at least not in my environment. With doctests, do you mean blocks of ``.. testcode::`` and ``.. testoutput::`` that are visible in the HOWTO?

**Comment by matthias-wright** [comment]
I only tested the ``train_step`` function for now because testing the ``train_epoch`` function takes around 2 minutes (on my machine).

File: docs/howtos/lr_schedule.rst [hunk 6]
```diff
@@ -1,138 +1,230 @@
...
+  train_ds_size = config.train_ds_size
+  steps_per_epoch = train_ds_size // config.batch_size
+  learning_rate_fn = create_learning_rate_fn(config, config.learning_rate, steps_per_epoch)
+  
+  rng = jax.random.PRNGKey(0)
+  state = create_train_state(rng, config, learning_rate_fn)
+  rng, _ = jax.random.split(rng)
+
+  batch = get_dummy_batch(config.batch_size)
+  state, metrics = train_step(state, batch, learning_rate_fn)
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:225]: Why not set `config.train_ds_size=128` and then call `train_epoch()` instead? That shouldn't take significantly longer and would test all your code.

**Review by andsteing — CHANGES_REQUESTED** [review]
Looks good, thanks for the updates!

I think you could indeed test `train_epoch()` (see file comment) and then the PR would be good to go.

File: docs/howtos/lr_schedule.rst [hunk 7]
```diff
@@ -1,138 +1,230 @@
...
+    """Creates initial `TrainState`."""
+    cnn = CNN()
+    params = cnn.init(rng, jnp.ones([1, 28, 28, 1]))['params']
+    tx = optax.sgd(learning_rate_fn, config.momentum) #!
+    return train_state.TrainState.create(
+        apply_fn=cnn.apply, params=params, tx=tx)
+
+
+
+.. testcode::
```
Comment (andsteing) [inline docs/howtos/lr_schedule.rst:212]: I think if you used `testsetup` here then it would still test it but not show it in the HOWTO, which would be preferable.

**Comment by matthias-wright** [comment]
Thanks for the feedback! I had to use ``testcleanup`` instead of ``testsetup`` but the result is the same. The tests are executed during build but aren't visible in the HOWTO.

**Review by andsteing — APPROVED** [review]
Thanks for updating the HOWTO!
It's great to have it properly tested now...
